dmscripts/compare_dmp_and_dnb_data.py

```diff
@@ -0,0 +1,324 @@
+        # iterate all the suppliers of the frameworks
+        for framework in frameworks:
+            dmp_response = self.get_dmp_supplier_data(framework=framework, from_declaration=from_declaration)
+            if from_declaration:
+                dmp_response = dmp_response.get('supplierFrameworks', [])
+            else:
+                dmp_response = dmp_response.get('suppliers', [])
+            for dmp_supplier in dmp_response:
+                duns_number = dmp_supplier.get('declaration', {}).get('supplierDunsNumber') if from_declaration else dmp_supplier.get('dunsNumber')
+                # don't compare if DMp has no DUNS for a supplier
+                if duns_number is None:
+                    print('NO DUNS NUMBER FOR', self.extract_dmp_supplier_data(dmp_supplier, from_declaration=from_declaration))  # TODO: raise/log error?
+                    continue
+                # excluding existing supplier of the CSV write or print the rest
+                if duns_number not in self.duns_number_compared:
+                    dmp_data = self.extract_dmp_supplier_data(dmp_supplier, from_declaration=from_declaration)
+                    try:
+                        dnb_data = self.extract_dnb_org_data(self.get_dnb_org_data(duns_number))
+                        row = self.compare_data(
+                            duns_number,
+                            dmp_data,
+                            dnb_data
+                        )
+                        if export_dnb_data:
+                            row += dnb_data
+                        if export_dmp_data:
+                            row += dmp_data
+                    except DNBAPIException as e:
+                        row = [
+                            duns_number,
+                            datetime.datetime.now(),  # timestamp
+                            0.,  # is in D&B
+                            0.,  # company's name
+                            0.,  # company house
+                            0.,  # country exact match
+                            0.,  # postcode exact match
+                            0.,  # company's address
+                            f'{e.code}: {e.text}',  # D&B failure reason
+                        ]
+                        if export_dnb_data:
+                            row += [None] * 5
+                        if export_dmp_data:
+                            row += dmp_data
+                    if csv_filename is not None:
+                        counter += 1
+                        print(duns_number)
+                        writer.writerow(row)  # TODO: extend with D&B data
+                        self.duns_number_compared.add(duns_number)  # add DUNS to set
+                        if counter >= self.max_dnb_api_calls_per_method:
+                            return
+                    else:
+                        print(row)  # TODO: log maybe?
+        # close file resource if necessary
```

Comment:
ideally this would be done with a `finally` block or context manager, but good enough for a prototype script...

Reply:
> ideally this would be done with a `finally` block or context manager, but good enough for a prototype script...

Agreed, but as you said it is overkill for a prototype.